service/evaluation/evaluation.go

```diff
@@ -0,0 +1,845 @@
+		if control.ParentControlId != nil {
+			continue
+		}
+
+		// Collect control including sub-controls from the controls_in_scope slice to start them later as scheduler jobs
+		controls = append(controls, control)
+		controls = append(controls, control.GetControls()...)
+
+		// Check if scheduler job for current control_id is already running
+		jobs, err := s.scheduler.FindJobsByTag(schedulerTag)
+		if len(jobs) > 0 && err == nil {
+			err = fmt.Errorf("evaluation for Cloud Service '%s' and Control ID '%s' already started", toe.GetCloudServiceId(), control.GetId())
+			log.Error(err)
+			return nil, status.Errorf(codes.AlreadyExists, "%s", err)
+		}
+
+		// parentSchedulerTag is the tag for the parent control (e.g., OPS-13)
+		parentSchedulerTag = getSchedulerTag(toe.GetCloudServiceId(), control.GetId())
+
+		// Add number of sub-controls to the WaitGroup. For the control (e.g., OPS-13) we have to wait until all the sub-controls (e.g., OPS-13.1) are ready.
+		// Control is a parent control if no parentControlId exists.
+		s.wg[parentSchedulerTag] = &WaitGroup{
+			wg:      &sync.WaitGroup{},
+			wgMutex: sync.Mutex{},
+		}
+		s.wg[parentSchedulerTag].wgMutex.Lock()
+		// The controls list contains also the parent control itself and must be minimized by 1 for the parent_control_id.
+		s.wg[parentSchedulerTag].wg.Add(len(controls) - 1)
+		s.wg[parentSchedulerTag].wgMutex.Unlock()
+
+		// Add control including sub-controls to the scheduler
+		for _, control := range controls {
+			// If the control is a sub-control create parentSchedulerTag
+			if control.GetParentControlId() == "" {
+				parentSchedulerTag = ""
+			} else {
+				// parentSchedulerTag is the tag for the parent control (e.g., OPS-13)
+				parentSchedulerTag = getSchedulerTag(toe.GetCloudServiceId(), control.GetParentControlId())
+			}
+			err = s.addJobToScheduler(control, toe, parentSchedulerTag, interval)
+			// We can return the error as it is
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		log.Infof("Evaluate Control ID '%s' for Cloud Service '%s' every 5 minutes...", control.GetId(), toe.GetCloudServiceId())
+	}
+
+	// Start scheduler jobs
+	s.scheduler.StartAsync()
+	log.Infof("Evaluation started.")
+
+	resp = &evaluation.StartEvaluationResponse{}
+
+	return
+}
+
+// StopEvaluation is a method implementation of the evaluation interface: It stops the evaluation for a TargetOfEvaluation.
+func (s *Service) StopEvaluation(_ context.Context, req *evaluation.StopEvaluationRequest) (resp *evaluation.StopEvaluationResponse, err error) {
+
+	var (
+		schedulerTags []string
+		// control      *orchestrator.Control
+		toe *orchestrator.TargetOfEvaluation
+	)
+
+	// Validate request
+	err = service.ValidateRequest(req)
+	if err != nil {
+		err = status.Errorf(codes.InvalidArgument, "%v", err)
+		return
+	}
+
+	// Get Target of Evaluation. The Target of Evaluation is retrieved to get the controls_in_scope, which are then stopped from the evaluation.
```

Comment:
Wouldn't it be easier to have a tag for the ToE itself (cloudservice+catalog) instead of looping over all controls again? Seems fragile.

Reply:
See the answer above. Now I am adding a scheduler for each ToE and remove it when stopping the evaluation for the specific ToE.